src/main/java/ch/naviqore/raptor/router/Query.java

```diff
@@ -90,17 +91,66 @@ class Query {
      * </ul>
      */
     List<StopLabelsAndTimes.Label[]> run() {
-        // set up footpath relaxer and route scanner and inject stop labels and times
-        FootpathRelaxer footpathRelaxer = new FootpathRelaxer(stopLabelsAndTimes, raptorData,
-                config.getMinimumTransferDuration(), config.getMaximumWalkingDuration(), timeType);
-        RouteScanner routeScanner = new RouteScanner(stopLabelsAndTimes, raptorData,
-                config.getMinimumTransferDuration(), timeType, referenceDate, maxDaysToScan);
 
         // initially relax all source stops and add the newly improved stops by relaxation to the marked stops
         Set<Integer> markedStops = initialize();
         markedStops.addAll(footpathRelaxer.relaxInitial(sourceStopIndices));
         markedStops = removeSuboptimalLabelsForRound(0, markedStops);
 
+        // if range is 0 or smaller there is no range, and we don't need to rerun rounds with different start offsets
+        if (raptorRange <= 0) {
+            doRounds(markedStops);
+        } else {
+            doRangeRaptor(markedStops);
+        }
+        return stopLabelsAndTimes.getBestLabelsPerRound();
+    }
+
+    void doRangeRaptor(Set<Integer> markedStops) {
+        // prepare range offsets
+        List<Integer> rangeOffsets = getRangeOffsets(markedStops, routeScanner);
+        HashMap<Integer, Integer> stopIdxSourceTimes = new HashMap<>();
+        for (int stopIdx : markedStops) {
+            stopIdxSourceTimes.put(stopIdx, stopLabelsAndTimes.getLabel(0, stopIdx).targetTime());
+        }
+
+        // scan all range offsets in reverse order (earliest arrival / latest departure first)
+        for (int offsetIdx = rangeOffsets.size() - 1; offsetIdx >= 0; offsetIdx--) {
+            int rangeOffset = rangeOffsets.get(offsetIdx);
+            int timeFactor = timeType == TimeType.DEPARTURE ? 1 : -1;
+            log.debug("Running rounds with range offset {}", rangeOffset);
+
+            // set source times to the source times of the previous round
```

Comment:
We need to reset them because the targetTime of the initial label has to be updated for each range offset.
So if our range raptor spawns for minutes 0, 15, 30, we want to set the Initial Label Target time to departure time + 30 minutes, then departure time + 15 minutes and finally departure time.